### PR TITLE
Add anonymized json plan in QueryCompletedEvent

### PR DESCRIPTION
**Pull request by gaurav8297**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

This change will eventually help us collect query plans from trino (or products built on trino) users without leaking any sensitive data. These plans then can be used to drive analytics around how people are using trino which essentially helps in improving the product.

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 1]
```diff
@@ -0,0 +1,785 @@
...
+import static io.trino.sql.planner.plan.WindowNode.Specification;
+import static java.util.Map.entry;
+import static java.util.Objects.requireNonNull;
+
+public class AnonymizePlanRewriter
+        extends SimplePlanRewriter<Void>
+{
+    private final Metadata metadata;
+    private final Session session;
+    private final Map<String, String> anonymizedSymbols = new HashMap<>();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:107]: Could you try re-using `SymbolMapper` for anonymising Symbols ?

Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:107]: > Could you try re-using `SymbolMapper` for anonymising Symbols ?

I thought of using `SymbolMapper` but we need to anonymize other things too like expressions, table handle, column handle etc.

File: core/trino-main/src/main/java/io/trino/event/QueryMonitor.java [hunk 2]
```diff
@@ -346,6 +355,30 @@ private Optional<String> createTextQueryPlan(QueryInfo queryInfo)
         return Optional.empty();
     }
 
+    private Optional<String> createAnonymizedPlan(QueryInfo queryInfo)
+    {
+        try {
+            if (queryInfo.getOutputStage().isPresent()) {
+                Map<PlanFragmentId, PlanNode> anonymizedPlan = new HashMap<>();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:362]:
```
Map<PlanFragmentId, PlanNode> anonymizedPlan = getAllStages(queryInfo.getOutputStage()).stream()
                        .map(StageInfo::getPlan)
                        .filter(Objects::nonNull)
                        .collect(toImmutableMap(PlanFragment::getId, plan -> rewriteWith(rewriter, plan.getRoot())));
```

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 3]
```diff
@@ -0,0 +1,785 @@
...
+                handle.getCatalogName(),
+                new ConnectorTransactionHandle() {},
+                new ConnectorIndexHandle() {});
+    }
+
+    private TableHandle anonymize(TableHandle handle)
+    {
+        TableSchema tableSchema = metadata.getTableSchema(session, handle);
+        return new TableHandle(
+                handle.getCatalogName(),
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:642]: We need to annonymize catalog name here

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+
+            anonymizedSymbols.put(anonymize(key), anonymizedValue);
+        });
+        return anonymizedSymbols.build();
+    }
+
+    private IndexHandle anonymize(IndexHandle handle)
+    {
+        return new IndexHandle(
+                handle.getCatalogName(),
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:633]: We need to annonymize catalog name here

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+    {
+        TableSchema tableSchema = metadata.getTableSchema(session, handle);
+        return new TableHandle(
+                handle.getCatalogName(),
+                new AnonymizedTableHandle(
+                        new QualifiedObjectName(
+                                anonymize(tableSchema.getQualifiedName().getCatalogName()),
+                                anonymize(tableSchema.getQualifiedName().getSchemaName()),
+                                anonymize(tableSchema.getQualifiedName().getObjectName())),
+                        handle.getConnectorHandle().getClass().getName()),
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:648]: You can get connector name by looking it up in output of `Metadata#getCatalogs` instead of class name of connector handle

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 6]
```diff
@@ -0,0 +1,785 @@
...
+        return new RefreshMaterializedViewNode(
+                node.getId(),
+                new QualifiedObjectName(
+                        anonymize(node.getViewName().getCatalogName()),
+                        anonymize(node.getViewName().getSchemaName()),
+                        anonymize(node.getViewName().getObjectName())));
+    }
+
+    @Override
+    public PlanNode visitTableWriter(TableWriterNode node, RewriteContext<Void> context)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:377]: It looks like there is information needing anonymisation in TableWriterNode as well

File: core/trino-main/src/main/java/io/trino/event/QueryMonitor.java [hunk 7]
```diff
@@ -346,6 +355,30 @@ private Optional<String> createTextQueryPlan(QueryInfo queryInfo)
...
+                AnonymizePlanRewriter rewriter = new AnonymizePlanRewriter(metadata, queryInfo.getSession().toSession(sessionPropertyManager));
+                List<StageInfo> allStages = getAllStages(queryInfo.getOutputStage());
+                for (StageInfo stageInfo : allStages) {
+                    if (stageInfo.getPlan() == null) {
+                        continue;
+                    }
+                    anonymizedPlan.put(stageInfo.getPlan().getId(), rewriteWith(rewriter, stageInfo.getPlan().getRoot()));
+                }
+                if (!anonymizedPlan.isEmpty()) {
+                    return Optional.of(anonymizedPlanCodec.toJson(anonymizedPlan));
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:372]: One issue here is that we're relying on AnonymizePlanRewriter to anonymise the PlanNode and then printing the whole PlanNode. It's very easy to miss updating AnonymizePlanRewriter after some new info is added to a PlanNode. 
So maybe we should rather have `AnonymizePlanRewriter` itself produce the anonymized plan fragment string and we can explicitly print the fields that we need there instead of relying on json serialisation of PlanNode. `JoinOrderPrinter` is also an example of similar approach.

File: core/trino-main/src/main/java/io/trino/event/QueryMonitor.java [hunk 8]
```diff
@@ -249,6 +257,7 @@ private QueryMetadata createQueryMetadata(QueryInfo queryInfo)
                 queryInfo.getRoutines(),
                 queryInfo.getSelf(),
                 createTextQueryPlan(queryInfo),
+                createAnonymizedPlan(queryInfo),
```
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: We already send the full plan with all details via `createTextQueryPlan` on above line in the query completion event.
QueryContext also has stuff like user and principal. So I don't think adding the anonymised plan is going to expose any new info to the query completion events.
If we must introduce a toggle for anonymisation of plan sent in query completion event, then it would make more sense to send the anonymised plan in place of the current full text query plan.

[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: Yeah - true. I did not know we have a plan in QCE already :)

[the remark this answers is not in the document]
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: So why are adding new field for anonymized plan. It would make more sense (to me) to reuse the field we already have, and have a toggle which tell Trino, should normal or anonymized plan be recorded.

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: Or, anonymization done in the event consumer.

[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: > Or, anonymization done in the event consumer.

The problem with anonymization in the event consumer is that the plan we get there is in text format and it's very hard to parse that plan to anonymize symbols, expressions or literals. Also, extracting any useful information from that plan is also not easy through code. That's why we want this anonymised plan in JSON format so that we could easily parse it later on.

[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:260]: > why are adding new field for anonymized plan

It's possible that someone wants to record the full text plan and anonymised plan in different downstream systems to support different use cases. Currently we send the same event to all event listeners, so reusing the same field would take away some flexibility here.

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 9]
```diff
@@ -0,0 +1,785 @@
...
+    }
+
+    private Symbol anonymize(Symbol symbol)
+    {
+        return new Symbol(anonymize(symbol.getName()));
+    }
+
+    private String anonymize(String name)
+    {
+        return anonymizedSymbols.computeIfAbsent(name, key -> String.valueOf(name.hashCode()));
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:705]: Let's also add a prefix of what was anonymised (table, column, symbol, literal etc.) before the hash so that it's easier to read the anonymised output

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 10]
```diff
@@ -0,0 +1,785 @@
...
+                    @Override
+                    public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        return anonymize(Symbol.from(node)).toSymbolReference();
+                    }
+
+                    @Override
+                    public Expression rewriteLiteral(Literal node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        if (node instanceof StringLiteral) {
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:729]: I think we have to anonymise CharLiteral, GenericLiteral and BinaryLiteral

File: core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java [hunk 11]
```diff
@@ -0,0 +1,785 @@
...
+                    }
+                },
+                expression);
+    }
+
+    public static class AnonymizedTableHandle
+            implements ConnectorTableHandle
+    {
+        private final QualifiedObjectName tableName;
+        private final String className;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/AnonymizePlanRewriter.java:743]: Should be connectorName after https://github.com/trinodb/trino/pull/12345/files#r871282206

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/TableScanNodeRepresentation.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    private final String catalog;
+    private final String schema;
+    private final String table;
+    private final Map<String, Type> columns;
+
+    @JsonCreator
+    public TableScanNodeRepresentation(
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("name") String name,
+            @JsonProperty("sources") List<AnonymizedNodeRepresentation> sources,
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/TableScanNodeRepresentation.java:40]: I don't think table scan will have any sources

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ValuesNodeRepresentation.java [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        extends AnonymizedNodeRepresentation
+{
+    private final int rowCount;
+    private final Optional<List<String>> rows;
+
+    @JsonCreator
+    public ValuesNodeRepresentation(
+            @JsonProperty("id") PlanNodeId id,
+            @JsonProperty("name") String name,
+            @JsonProperty("sources") List<AnonymizedNodeRepresentation> sources,
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ValuesNodeRepresentation.java:36]: ValuesNode shouldn't have any sources

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+        COLUMN,
+        SYMBOL,
+        LITERAL
+    }
+
+    private final Metadata metadata;
+    private final Session session;
+    private final JsonCodec<AnonymizedNodeRepresentation> representationJsonCodec = jsonCodec(AnonymizedNodeRepresentation.class);
+
+    private final Map<PlanNodeId, AnonymizedNodeRepresentation> representations = new HashMap<>();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:83]: This map and `addRepresentation` is unnecessary since you're just serialising from the top most node recursively.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+    {
+        node.accept(new Visitor(), null);
+        return representationJsonCodec.toJson(representations.get(node.getId()));
+    }
+
+    private class Visitor
+            extends PlanVisitor<Void, Void>
+    {
+        @Override
+        protected Void visitPlan(PlanNode node, Void context)
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:101]: Return `AnonymizedNodeRepresentation` instead of `Void`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+                            .collect(toImmutableMap(
+                                    entry -> anonymize(entry.getKey()),
+                                    entry -> anonymize(unresolveFunctions(entry.getValue())).toString()))));
+            return context;
+        }
+
+        @Override
+        public Void visitSort(SortNode node, Void context)
+        {
+            addRepresentation(new SortNodeRepresentation(
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:255]: Let's also add TopNRepresentation

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ExchangeNodeRepresentation
+        extends AnonymizedNodeRepresentation
+{
+    private final ExchangeNode.Type type;
+    private final ExchangeNode.Scope scope;
+    private final PartitioningHandle partitioningHandle;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java:35]: We shouldn't rely on `PartitioningHandle` directly, the fields contained in it can change. I think we just need ConnectorPartitioningHandle from it but we need to anonymise it somehow as it can contain column names depending on what the connector does.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    private final List<String> partitionArguments;
+    private final Optional<Symbol> partitionHashColumn;
+    private final boolean replicateNullsAndAny;
+    private final Optional<OrderingScheme> orderingScheme;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java:39]: Don't rely on `OrderingScheme` directly, extract `orderings` from it

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/RemoteSourceNodeRepresentation.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class RemoteSourceNodeRepresentation
+        extends AnonymizedNodeRepresentation
+{
+    private final List<PlanFragmentId> sourceFragmentIds;
+    private final Optional<OrderingScheme> orderingScheme;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/RemoteSourceNodeRepresentation.java:34]: Don't rely on OrderingScheme directly, extract orderings from it

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/RemoteSourceNodeRepresentation.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    private final ExchangeNode.Type exchangeType; // This is needed to "unfragment" to compute stats correctly.
+    private final RetryPolicy retryPolicy;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/RemoteSourceNodeRepresentation.java:36]: We don't need `retryPolicy`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/SortNodeRepresentation.java [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+import io.trino.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class SortNodeRepresentation
+        extends AnonymizedNodeRepresentation
+{
+    private final OrderingScheme orderingScheme;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/SortNodeRepresentation.java:29]: Don't rely on OrderingScheme directly, extract orderings from it

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/TableScanNodeRepresentation.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+public class TableScanNodeRepresentation
+        extends AnonymizedNodeRepresentation
+{
+    private final String connector;
+    private final String catalog;
+    private final String schema;
+    private final String table;
+    private final Map<Symbol, String> assignments;
+
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/TableScanNodeRepresentation.java:35]: We should also record anonymised `enforcedConstraint`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/WindowNodeRepresentation.java [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Optional;
+import java.util.Set;
+
+import static io.trino.sql.planner.plan.WindowNode.Specification;
+import static java.util.Objects.requireNonNull;
+
+public class WindowNodeRepresentation
+        extends AnonymizedNodeRepresentation
+{
+    private final Specification specification;
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/WindowNodeRepresentation.java:33]: Don't rely on Specification from WindowNode directly, extract `partitionBy` and `orderings` from it

File: core/trino-main/src/test/java/io/trino/sql/planner/planprinter/TestAnonymizePlanPrinter.java [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+    public Object[][] planNodes()
+    {
+        return new Object[][] {
+                {
+                        buildPlan(pb -> pb.aggregation(ab -> ab
+                            .step(FINAL)
+                            .addAggregation(pb.symbol("sum", BIGINT), expression("sum(x)"), ImmutableList.of(BIGINT))
+                            .singleGroupingSet(pb.symbol("y", BIGINT), pb.symbol("z", BIGINT))
+                            .source(pb.values(pb.symbol("x", BIGINT), pb.symbol("y", BIGINT), pb.symbol("z", BIGINT))))),
+                        "{\n" +
```
[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/test/java/io/trino/sql/planner/planprinter/TestAnonymizePlanPrinter.java]: > compare the actual and expected node representation object

Idk if it makes sense to create the same object again in tests from PlanNode, it seems like with this kind of testing we are just making the future change harder. Also, I think if we have to, we should only test the API which returns the data in JSON format. WDYT? @raunaqmorarka

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java (hunk 17 above)
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java]: > I think we just need ConnectorPartitioningHandle from it but we need to anonymise it somehow as it can contain column names depending on what the connector does.

I think this is not possible at least in `main` because we can't cast it to connector level info.

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java]: We could extract what we can out of PartitioningHandle->ConnectorPartitioningHandle like the class name and the details if it is instance of SystemPartitioningHandle.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+        {
+            return WindowNodeRepresentation.fromPlanNode(node, typeProvider, getSourceRepresentations(node));
+        }
+
+        @Override
+        public AnonymizedNodeRepresentation visitTopN(TopNNode node, Void context)
+        {
+            return TopNNodeRepresentation.fromPlanNode(node, typeProvider, getSourceRepresentations(node));
+        }
+
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:175]: Let's also add DeleteNode, UpdateNode, TableWriterNode, LimitNode

File: core/trino-main/src/main/java/io/trino/execution/TableInfo.java [hunk 26]
```diff
@@ -21,23 +21,34 @@
...
 import javax.annotation.concurrent.Immutable;
 
+import java.util.Optional;
+
 import static java.util.Objects.requireNonNull;
 
 @Immutable
 public class TableInfo
 {
+    private final Optional<String> connectorName;
```
[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/execution/TableInfo.java:31]: The reason why we need it here is that to fetch `connectorName` the transaction should be open but by the time we reach anonymized plan generation the transaction is closed. Even if you look at PlanPrinter there also we are using TableInfo to display table level predicate.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.trino.sql.planner.plan.PlanNodeId;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+@JsonTypeInfo(
```
[the remark this answers is not in the document]
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java:26]: I've removed `name` from all classes except Default representation. Also please note that @type will be required for deserialization to specific implemenation.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+        @JsonSubTypes.Type(value = FilterNodeRepresentation.class, name = "filter"),
+        @JsonSubTypes.Type(value = WindowNodeRepresentation.class, name = "window"),
+        @JsonSubTypes.Type(value = TopNNodeRepresentation.class, name = "topN"),
+        @JsonSubTypes.Type(value = SortNodeRepresentation.class, name = "sort"),
+        @JsonSubTypes.Type(value = RemoteSourceNodeRepresentation.class, name = "remoteSource"),
+        @JsonSubTypes.Type(value = ExchangeNodeRepresentation.class, name = "exchange"),
+        @JsonSubTypes.Type(value = JoinNodeRepresentation.class, name = "join"),
+        @JsonSubTypes.Type(value = SemiJoinNodeRepresentation.class, name = "semiJoin"),
+})
+public abstract class AnonymizedNodeRepresentation
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java:44]: Why do we need a separate hierarchy for the anonymized representation? It should be able to encode the same information in a regular plan representation, with all sensitive information masked or removed

**Review by martint — COMMENTED** [review]
Please make sure this fits in with the changes being made here: https://github.com/trinodb/trino/pull/9940

Ideally, this PR should be rebased on top of that one to minimize conflicts.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java (hunk 28 above)
Comment (gaurav8297) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java:44]: There are few problems wrt simply anonymizing each `PlanNode` and expose the tree as JSON.

1. In case, any new field is added to any `PlanNode`, which contains some sensitive information, we could leak that information. So, we would always have to remember to update this code whenever any plan node changes.
2. Another issue is that it seems like we can't serialize `PlanNode` into JSON because it contains different expressions which don't implement Jackson annotations.

And if we go with the `jsonFragmentPlan` route in the `PlanPrinter`, which the above PR kind of followed. Then, the issue is that the result JSON contains all the specific plan node fields under the `details` key in a text format that is hard to parse and subject to change.

For instance:
```
{
    "id" : "876",
    "name" : "InnerJoin",
    "identifier" : "[(\"orderkey_7\" = \"orderkey_4\")][$hashvalue_38, $hashvalue_39]",
    "details" : "Distribution: REPLICATED\ndynamicFilterAssignments = {orderkey_4 -> #df_1052}",
    ...
}
```

cc @raunaqmorarka @sopel39

Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java:44]: I'm not talking about anonymizing PlanNodes. That's the wrong approach. However, we could have a converter from PlanNode -> PlanRepresentation that is designed specifically to produce anonymized information. It would not suffer from the issue you describe because it would be a "whitelist" of conversions.

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizedNodeRepresentation.java:44]: `NodeRepresentation` in `PlanRepresentation` forces all the node level details into a single `details` string that isn't guaranteed to be in any particular format. For our use case we want the details to be easily parseable by some downstream consumer of event listener.
I suppose we could stuff all the anonymized details into a map and convert that into a json string. But that seems an unnatural approach compared to serialization of representation objects we've done here.
It's also odd behaviour for `EXPLAIN (TYPE LOGICAL, FORMAT JSON)` to give up on json format and start printing unformatted lines for the `details` part of query plans.
How about we extend `NodeRepresentation` to be a class hierarchy like we have here ? That would allow `EXPLAIN (TYPE LOGICAL, FORMAT JSON)` to print json all the way and the plan anonymiser we have here would be able to reuse the same hierarchy to populate anonymized info.

File: core/trino-main/src/main/java/io/trino/execution/TableInfo.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/execution/TableInfo.java:31]: Okay, let's do it in a preparatory commit then.

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java (hunk 25 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:180]: ping

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.airlift.json.JsonCodec.jsonCodec;
+import static io.trino.execution.StageInfo.getAllStages;
+import static io.trino.sql.planner.TypeProvider.getTypeProvider;
+import static java.util.Objects.requireNonNull;
+
+public class AnonymizePlanPrinter
+{
+    private static final JsonCodec<Map<PlanFragmentId, AnonymizedNodeRepresentation>> ANONYMIZED_PLAN_JSON_CODEC = jsonCodec(new TypeToken<>() {});
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/AnonymizePlanPrinter.java:57]: `mapJsonCodec(PlanFragmentId.class, AnonymizedNodeRepresentation.class)`

File: core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java (hunk 17 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/sql/planner/planprinter/anonymize/ExchangeNodeRepresentation.java]: ping
